### PR TITLE
Add gru component

### PR DESCRIPTION
**Pull request by yuuee-www**
# What does this PR do?

This PR introduces the integration of the Gradient Rectified Unlearning (GRU) with the Gradient Ascent (GA) unlearning objective. Currently, we are providing the GRU with the GA version due to time constraints. Integrations with other unlearning objectives, such as NPO, SimNPO, WGA, and GD, are planned and will be added in subsequent updates as soon as possible. For more detailed information, please refer to the original paper on [arXiv](https://www.arxiv.org/abs/2503.09117).

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Have you gone through the documentation of adding new [components](../docs/components.md)?
- [x] Did you make sure to update the documentation with your changes? Here are the pointers to documentation
      [documentation guidelines](../README.md#-further-documentation).

## Sample Implementation and Some Reproduction Results
We run some examples following the common hyperparameters in [docs/evaluation.md](docs/evaluation.md) for reproducibility purposes. As DeepSpeed does not support gradient manipulation currently, it is not used in our implementation. An example command for executing the GRU on forget10 is as follows::
```bash
CUDA_VISIBLE_DEVICES=0 python src/train.py --config-name=unlearn.yaml experiment=unlearn/tofu/default forget_split=forget10 retain_split=retain90 trainer=GRU task_name=gru_forget10
```
The reproducible results on the Llama-3.2-1B-Instruct architecture:

### TOFU unlearning on the `Llama-3.2-1B-Instruct` architecture

<div style="overflow-x: auto; max-width: 100%;">
<table class="dataframe">
  <thead>
    <tr>
      <th>Method</th>
      <th style="text-align: center;" colspan="3" halign="left">forget01</th>
      <th style="text-align: center;" colspan="3" halign="left">forget05</th>
      <th style="text-align: center;" colspan="3" halign="left">forget10</th>
    </tr>
    <tr>
      <th></th>
      <th>forget_quality</th>
      <th>model_utility</th>
      <th>forget_truth_ratio</th>
      <th>forget_quality</th>
      <th>model_utility</th>
      <th>forget_truth_ratio</th>
      <th>forget_quality</th>
      <th>model_utility</th>
      <th>forget_truth_ratio</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Finetuned</th>
      <td>0.01</td>
      <td>0.60</td>
      <td>0.47</td>
      <td>2.96e-13</td>
      <td>0.6</td>
      <td>0.47</td>
      <td>8.08e-22</td>
      <td>0.6</td>
      <td>0.48</td>
    </tr>
    <tr>
      <th>Retain</th>
      <td>1.0</td>
      <td>0.60</td>
      <td>0.65</td>
      <td>1.0</td>
      <td>0.6</td>
      <td>0.63</td>
      <td>1.0</td>
      <td>0.59</td>
      <td>0.63</td>
    </tr>
    <tr>
      <td colspan="20"> </td>
    </tr>
    <tr>
      <th>GradAscent</th>
      <td>0.27</td>
      <td>0.33</td>
      <td>0.59</td>
      <td>1.94e-119</td>
      <td>0</td>
      <td>2.52e-23</td>
      <td>1.06e-239</td>
      <td>0</td>
      <td>2.25e-18</td>
    </tr>
    <tr>
      <th>GradDiff</th>
      <td>0.77</td>
      <td>0.43</td>
      <td>0.57</td>
      <td>1.94e-119</td>
      <td>0.53</td>
      <td>3.87e-34</td>
      <td>1.06e-239</td>
      <td>0.49</td>
      <td>3.53e-27</td>
    </tr>
    <tr>
      <th>IdkDPO</th>
      <td>0.01</td>
      <td>0.51</td>
      <td>0.60</td>
      <td>1.12e-05</td>
      <td>0.07</td>
      <td>0.62</td>
      <td>4.64e-12</td>
      <td>0.23</td>
      <td>0.6</td>
    </tr>
    <tr>
      <th>NPO</th>
      <td>0.92</td>
      <td>0.56</td>
      <td>0.66</td>
      <td>0.14</td>
      <td>0.45</td>
      <td>0.7</td>
      <td>0.02</td>
      <td>0.46</td>
      <td>0.7</td>
    </tr>
    <tr>
      <th>SimNPO</th>
      <td>0.58</td>
      <td>0.46</td>
      <td>0.55</td>
      <td>5.01e-100</td>
      <td>0.58</td>
      <td>4.19e-03</td>
      <td>2.47e-203</td>
      <td>0.54</td>
      <td>1.07e-05</td>
    </tr>
     <tr>
      <th>RMU</th>
      <td>0.16</td>
      <td>0.55</td>
      <td>0.70</td>
      <td>4.87e-10</td>
      <td>0.58</td>
      <td>0.77</td>
      <td>3.15e-15</td>
      <td>0.59</td>
      <td>0.76</td>
    </tr>
<tr>
      <th>GradAscent w/ GRU</th>
      <td>0.03</td>
      <td>0.59</td>
      <td>0.50</td>
      <td>1.39e-06</td>
      <td>0.49</td>
      <td>0.52</td>
      <td>6.83e-09</td>
      <td>0.54</td>
      <td>0.55</td>
    </tr>
  </tbody>
</table>
</div>

### CODE REVIEW

File: src/trainer/unlearn/gru.py [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+
+        self.gradient_accum = {}
+                        
+        self.memory_grad = {}
+
+    def orthogonal_component(self, g, g1):
+
+        g1g1 = self.compute_total_gradient_dot_product(g1, self.structure_map, g1, self.structure_map)
+        gg1 = self.dotp_retain
+        print(gg1/g1g1)
```
Comment (molereddy) [inline src/trainer/unlearn/gru.py:87]: can you clean up the code - e.g. remove this print and any other unnecessary prints/comments from the code?

Comment (yuuee-www) [reply src/trainer/unlearn/gru.py]: Thanks for the feedback! I've removed unnecessary print statements. Let me know if anything else needs updating!

**Comment by 99starman-ctrl** [comment]
Hi!

Sorry for jumping into the conversation, and really interesting method by the way! I tried to reproduce the TOFU forget10 results of GRU on the Llama-3.2-1B-Instruct, following the example command you had above, but the result doesn't seem to match. I ran twice but both time the `model_utility` is about 0.364 (table shows 0.54), and `forget_truth_ratio` is about 0.512 (table shows 0.55). I also ran `src/eval.py` following the same [default settings](https://github.com/locuslab/open-unlearning/blob/main/scripts/tofu_unlearn.sh#L62), and the two metrics gave the same results as above.

I also ran the same configurations with RMU, and its `model_utility` is about 0.574, which is close to the table (0.59), but the `forget_truth_ratio` is 0.588, which is actually worse. 

So I'm just wondering what your exact evaluation settings were and if you could double-check the settings, especially regarding model utility, which seems to be the main argument for GRU.

**Comment by yuuee-www** [comment]
> Hi!
> 
> Sorry for jumping into the conversation, and really interesting method by the way! I tried to reproduce the TOFU forget10 results of GRU on the Llama-3.2-1B-Instruct, following the example command you had above, but the result doesn't seem to match. I ran twice but both time the `model_utility` is about 0.364 (table shows 0.54), and `forget_truth_ratio` is about 0.512 (table shows 0.55). I also ran `src/eval.py` following the same [default settings](https://github.com/locuslab/open-unlearning/blob/main/scripts/tofu_unlearn.sh#L62), and the two metrics gave the same results as above.
> 
> I also ran the same configurations with RMU, and its `model_utility` is about 0.574, which is close to the table (0.59), but the `forget_truth_ratio` is 0.588, which is actually worse.
> 
> So I'm just wondering what your exact evaluation settings were and if you could double-check the settings, especially regarding model utility, which seems to be the main argument for GRU.

Thank you for your interest in our method and for taking the time to test it out! We appreciate your efforts. Let me address your points:

We’ve uploaded an eval output JSON log file
[TOFU_EVAL.json](https://github.com/user-attachments/files/19475820/TOFU_EVAL.json)
 that matches the results reported in our table (0.54 for model_utility and 0.55 for forget_truth_ratio). If possible, could you share the specific evaluation and unlearning settings you used or even upload your own eval JSON log file? This would help us figure out where the issue might lie.

Regarding the RMU results you mentioned, we’d like to kindly clarify that the numbers in the table were directly taken from the original table in [results.md](https://github.com/locuslab/open-unlearning/blob/main/docs/results.md). We didn’t conduct any additional evaluations for RMU ourselves. The only row we added to the table was for GradAscent w/ GRU, which reflects the results from our work.


Finally, we want to note that the results in our table were produced using the common hyperparameters provided by the open-unlearning authors in [results.md](https://github.com/locuslab/open-unlearning/blob/main/docs/results.md). These were used for reproducibility purposes, and the table isn’t intended for direct comparison between methods, as the authors themselves have stated.

Once again, we truly appreciate your interest in our work and your efforts to test it. Let us know if there's anything more we can assist with!

**Comment by rtx-1999** [comment]
> 
> Hi @rtx-1999,
> 
> Thank you for your interest in our method and for taking the time to test it out! We appreciate your efforts. Let me address your points:
> 
> We’ve uploaded an eval output JSON log file [TOFU_EVAL.json](https://github.com/user-attachments/files/19475820/TOFU_EVAL.json) that matches the results reported in our table (0.54 for model_utility and 0.55 for forget_truth_ratio). If possible, could you share the specific evaluation and unlearning settings you used or even upload your own eval JSON log file? This would help us figure out where the issue might lie.
> 
> Regarding the RMU results you mentioned, we’d like to kindly clarify that the numbers in the table were directly taken from the original table in [results.md](https://github.com/locuslab/open-unlearning/blob/main/docs/results.md). We didn’t conduct any additional evaluations for RMU ourselves. The only row we added to the table was for GradAscent w/ GRU, which reflects the results from our work.
> 
> Finally, we want to note that the results in our table were produced using the common hyperparameters provided by the open-unlearning authors in [results.md](https://github.com/locuslab/open-unlearning/blob/main/docs/results.md). These were used for reproducibility purposes, and the table isn’t intended for direct comparison between methods, as the authors themselves have stated.
> 
> Once again, we truly appreciate your interest in our work and your efforts to test it. Let us know if there's anything more we can assist with!


Thanks for your response and clarification on RMU results! Just wanted to clarify that making direct comparison between methods wasn't my purpose. I'm just trying to make sure that I have the same evaluation setup as yours. 

Here are the hydra configs, my evaluation script, eval.log, TOFU_EVAL.json, GRU.log and trainer_state.json, zipped in [settings.zip](https://github.com/user-attachments/files/19476493/settings.zip) as the yaml format is not supported for attachment here. As I checked, I used the exact setting for unlearning:
```
CUDA_VISIBLE_DEVICES=0 python src/train.py --config-name=unlearn.yaml experiment=unlearn/tofu/default forget_split=forget10 retain_split=retain90 trainer=GRU task_name=GRU_forget10
```
And for the evaluation script, I followed the default configs in [tofu_unlearn.sh](https://github.com/locuslab/open-unlearning/blob/main/scripts/tofu_unlearn.sh#L62), equivalently:
 
```
CUDA_VISIBLE_DEVICES=0 python src/eval.py \
experiment=eval/tofu/default.yaml \
forget_split=forget10 \
model=Llama-3.2-1B-Instruct \
task_name=GRU_forget10 \
model.model_args.pretrained_model_name_or_path=saves/unlearn/GRU_forget10 \
paths.output_dir=saves/unlearn/GRU_forget10/evals \
retain_logs_path=saves/eval/tofu_Llama-3.2-1B-Instruct_retain90/TOFU_EVAL.json
```

**Comment by Dornavineeth** [comment]
> CUDA_VISIBLE_DEVICES=0 python src/train.py --config-name=unlearn.yaml experiment=unlearn/tofu/default forget_split=forget10 retain_split=retain90 trainer=GRU task_name=GRU_forget10

@rtx-1999 I see that you are using only 1 device for unlearning. Using different number of GPU can effect the results quite a lot as the effective hyper parameters such as batch size etc are changed. 

Additionally, even if we manage to keep the effective hyper params same, using deepspeed can lead to [different results](https://mp.weixin.qq.com/s/hmIvSsZ2-yKEeiXUNRlTQw).

@yuuee-www can you confirm how many GPUs have you used for unlearning?

**Comment by Dornavineeth** [comment]
Hi @yuuee-www,

Thank you for taking the time to add your method to our repository. We've added [guidelines](https://github.com/locuslab/open-unlearning/blob/main/docs/contributing.md) on contributing to our repository, including how to share your settings, results, and reproducible scripts. Could you review them and update this PR accordingly?

This should help others use and reproduce your work in their studies.

Quick summary  of the contribution guide:

1. Create a folder under [community/methods/](https://github.com/locuslab/open-unlearning/tree/main/community/methods).
2. Include your method details, results, hyperparameter search scripts, and reproducibility steps.
3. Update the [leaderboard](https://github.com/locuslab/open-unlearning/blob/main/community/leaderboard.md) with your results.

We've provided a [template](https://github.com/locuslab/open-unlearning/tree/main/community/methods/template) folder to get you started.

Let us know if you have any questions!

**Comment by yuuee-www** [comment]
> Hi @yuuee-www,
> 
> Thank you for taking the time to add your method to our repository. We've added [guidelines](https://github.com/locuslab/open-unlearning/blob/main/docs/contributing.md) on contributing to our repository, including how to share your settings, results, and reproducible scripts. Could you review them and update this PR accordingly?
> 
> This should help others use and reproduce your work in their studies.
> 
> Quick summary of the contribution guide:
> 
> 1. Create a folder under [community/methods/](https://github.com/locuslab/open-unlearning/tree/main/community/methods).
> 2. Include your method details, results, hyperparameter search scripts, and reproducibility steps.
> 3. Update the [leaderboard](https://github.com/locuslab/open-unlearning/blob/main/community/leaderboard.md) with your results.
> 
> We've provided a [template](https://github.com/locuslab/open-unlearning/tree/main/community/methods/template) folder to get you started.
> 
> Let us know if you have any questions!

Hi @Dornavineeth,

Understood. We’ll follow the guidelines and update the PR soon.

Cc @rtx-1999.

Thanks!

**Comment by molereddy** [comment]
**A couple of requests:**
1. Can you mark out the trainer code so that it is clearer which of the code is your changes and which is from the HuggingFace source code? There are some comments indicating that, but it is not totally clear. Also, to the extent convenient, add comments that explain what is going on with your additions.
2. Given GRU can add onto many unlearning losses, it would be better if you could implement this in a more extensible way. You may introduce a `forget_loss_type` in the trainer config's `method_args` (like the `retain_loss_type`) so that users can vary the underlying forget loss with a config attribute. This can take values like NPO, DPO, GradAscent, SimNPO etc. You may use the trainer utils file to get these functionalities (and refer other trainer class implementations for their usage). In your GRU trainer you can now have `if` conditions on `forget_loss_type` that use different loss functions depending on the config.

For your leaderboard results, you may report your best-performing `forget_loss_type` with GRU.

If you have a better way of doing this, or face difficulties with this plan, do let us know.

**Comment by molereddy** [comment]
I see that ruff quality tests are failing here. Make sure to follow the instructions in https://github.com/locuslab/open-unlearning/blob/main/docs/contributing.md#create-a-pull-request to ensure you are in dev env and can apply ruff formatting.

File: src/trainer/unlearn/gru.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+class GRU(GradDiff,UnlearnTrainer):
+    def __init__(self,  gamma_gru=0.8, forget_loss_type="GradAscent", *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.gamma_gru = gamma_gru
+        self.forget_loss_type = forget_loss_type
+        self.gradient_accumulation_steps = kwargs["args"].gradient_accumulation_steps
+        if self.ref_model is None and self.forget_loss_type == "NPO":
+            self.ref_model = self._prepare_ref_model(self.model)
+            #self.ref_model = self.model.to(self.args.device)
```
Comment (molereddy) [inline src/trainer/unlearn/gru.py:20]: clean code up

File: src/trainer/unlearn/gru.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+        #assert len(structure_map1) == len(structure_map2), "Both gradient structures must contain the same number of elements."
+
+        total_dot_product = 0.0
+        index = 0
+
+        # Ensure both gradient tensors are on the same device
+        flattened_grads1 = flattened_grads1.to('cuda')
+        flattened_grads2 = flattened_grads2.to('cuda')
+
+        # for ((name1, shape1), (name2, shape2)) in zip(structure_map1, structure_map2):
```
Comment (molereddy) [inline src/trainer/unlearn/gru.py:132]: clean code up

File: src/trainer/unlearn/gru.py [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+                param.grad = grad_slice.to(param.device)
+
+            index += size  # Update index to the start of the next gradient slice
+
+        if index != self.flattened_gradient.numel():
+            raise ValueError("Total number of gradient elements does not match the length of the flattened gradient tensor.")
+       
+    def pipeline(self):
+        if self.dotp_retain < 0:
+            #print("dotp_retain:",self.dotp_retain)
```
Comment (molereddy) [inline src/trainer/unlearn/gru.py:182]: clean up

File: community/methods/GRU/README.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+## Computational Setup
+
+
+- **GPU Details**: NVIDIA A100 80GB
+- **GPU Count**: The code for our method currently supports single GPU execution. We plan to enhance the codebase in the future to support multi-GPU configurations.
+
+
+# Results
+
+To replicate your results, provide a `run.sh` script that contains all necessary commands to reproduce the final results. Ensure the script is well-documented.
```
Comment (molereddy) [inline community/methods/GRU/README.md:29]: you can 
- remove all the template instructions from this doc which we've written ("provide a concise summary", etc.)
- add your results here

**Comment by molereddy** [comment]
We can get this in once: 
- the few small nit comments are resolved 
- code is tested
- tests pass

**Comment by molereddy** [comment]
Also, we use DeepSpeed for distributed training in our code and experiments, and provide documentation with that in mind. Please test if your code works with DeepSpeed and if not (looking at the code it seems it may not), add a note to your documentation describing when your method works.

File: community/methods/GRU/README.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+Provide a concise summary of your method details and its contributions. Please avoid using images to keep the repository size manageable.
+
+# Setup
+
+Please include the experimental setup such as
+
+- [ ] **Hyperparameters & Search Space:** Specify key hyperparameters, their search ranges, number of trials etc.
+- [ ] **Computational Setup:** Mention the type and number of GPUs used.
+- [ ] **DeepSpeed Configuration:** If any modifications were made to the default DeepSpeed config, specify them here. (You may include the config as a code block.)
```
Comment (molereddy) [inline community/methods/GRU/README.md:16]: Also, we use DeepSpeed for distributed training in our code and experiments, and provide documentation with that in mind. Please test if your code works with DeepSpeed and if not (looking at the code it seems it may not), add a note to your documentation describing in what your method works.

If DeepSpeed is not supported, mention that
